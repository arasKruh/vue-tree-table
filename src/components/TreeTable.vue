<template>
  <div>
    <div class="row">
      <slot
        v-for="column in columns"
        :columnData="column"
        name="headerTemplate"
      >
        <TreeColumnHeader :column-data="column" />
      </slot>
    </div>

    <TreeBody
      :data="data"
      :columns="columnsOrder"
    >
      <template
        slot="rowTemplate"
        slot-scope="rowProps"
      >
        <slot
          name="rowTemplate"
          :defaultOrder="rowProps.defaultOrder"
          :rowData="rowProps.rowData"
        />
      </template>
    </TreeBody>
  </div>
</template>

<script>
  import TreeBody from './TreeBody'
  import TreeColumnHeader from './TreeColumnHeader'

  export default {
    name: 'TreeTable',
    components: {TreeBody, TreeColumnHeader},
    props: {
      data: {
        type: Array,
        default: () => { return {} }
      },
      columns: {
        type: Array,
        default: () => { return [] }
      }
    },
    computed: {
      columnsOrder(){
        return this.columns.map(column => column.id);
      }
    }
  }
</script>

<style scoped>
  .row{
    display: flex;
    flex-flow: row wrap;
    justify-content: center;
  }
</style>