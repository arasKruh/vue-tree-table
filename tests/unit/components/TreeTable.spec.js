import { mount } from '@vue/test-utils'
import TreeTable from '../../../src/components/TreeTable'
import TreeTableWithRowTemplate from './examples/TreeTableWithRowTemplate'
import TreeTableDefaultRow from './examples/TreeTableDefaultRow'

describe('TreeTable', () => {

  describe('default settings', () => {

    it('renders data in order specified in columns', () => {
      const wrapper = mount(TreeTableDefaultRow, {
        propsData: {
          tableData: [{name: 'Ziuta', surname: 'Kozak' }, {name: 'Koziolek', surname: 'Matolek'}],
          columns: [{label: 'Surname', id: 'surname'}, {label: 'Name', id: 'name'}]
        }
      })
      const names = wrapper.text().replace(/(\r\n|\n|\r)/gm, "").split(" ").filter(Boolean)
      expect(names).toEqual(['Surname', 'Name', 'Kozak', 'Ziuta', 'Matolek', 'Koziolek'])
    })

  })

  describe('when row template provided', () => {

    it('first puts surname and then name', () => {
      const wrapper = mount(TreeTableWithRowTemplate, {
        propsData: {
          tableData: [{name: 'Ziuta', surname: 'Kozak' }, {name: 'Koziolek', surname: 'Matolek'}],
          columns: [{label: 'Name', id: 'name'}, {label: 'Surname', id: 'surname'}]
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('when row template not provided', () => {

    it('uses default row component', () => {
      const wrapper = mount(TreeTableDefaultRow, {
        propsData: {
          tableData: [{name: 'Ziuta', surname: 'Kozak' }, {name: 'Koziolek', surname: 'Matolek'}],
          columns: [{label: 'Name', id: 'name'}, {label: 'Surname', id: 'surname'}]
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})