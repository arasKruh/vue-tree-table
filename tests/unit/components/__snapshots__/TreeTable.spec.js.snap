// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeTable when row template not provided uses default row component 1`] = `
<div
  class="table"
>
  <div
    class="row"
  >
    <div
      class="cell"
    >
      
  Name

    </div>
    <div
      class="cell"
    >
      
  Surname

    </div>
  </div>
   
  <div>
    <div
      class="row"
    >
      <div
        class="cell"
      >
        
    Ziuta
  
      </div>
      <div
        class="cell"
      >
        
    Kozak
  
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="cell"
      >
        
    Koziolek
  
      </div>
      <div
        class="cell"
      >
        
    Matolek
  
      </div>
    </div>
  </div>
</div>
`;

exports[`TreeTable when row template provided first puts surname and then name 1`] = `
<div
  class="table"
>
  <div
    class="row"
  >
    <div
      class="cell"
    >
      
  Name

    </div>
    <div
      class="cell"
    >
      
  Surname

    </div>
  </div>
   
  <div>
    <div
      class="row tree-row--red"
    >
      <div
        class="cell"
      >
        
    Ziuta
  
      </div>
      <div
        class="cell"
      >
        
    Kozak
  
      </div>
    </div>
    <div
      class="row tree-row--red"
    >
      <div
        class="cell"
      >
        
    Koziolek
  
      </div>
      <div
        class="cell"
      >
        
    Matolek
  
      </div>
    </div>
  </div>
</div>
`;
